godot-core/src/builtin/vector_macros.rs

```diff
@@ -0,0 +1,221 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+#![macro_use]
+
+/// Implements a single unary operator for a vector type. Only used for `Neg` at the moment.
+macro_rules! impl_vector_unary_operator {
+    (
+        // Name of the vector type.
+        $vector:ty,
+        // Type of each individual component, for example `i32`.
+        $scalar:ty,
+        // Names of the components, with parentheses, for example `(x, y)`.
+        ($($components:ident),*),
+        // Name of the operator trait, for example `Neg`.
+        $operator:ident,
+        // Name of the function on the operator trait, for example `neg`.
+        $func:ident$(,)?
+    ) => {
+        // -vector
+        impl std::ops::$operator for $vector {
+            type Output = Self;
+            fn $func(mut self) -> Self::Output {
+                $(
+                    self.$components = self.$components.$func();
+                )*
+                self
+            }
+        }
+    }
+}
+
+/// Implements a single infix binary operator for a vector type, between two vectors and between a
+/// vector and a scalar in both directions.
+macro_rules! impl_vector_binary_operator {
+    (
+        // Name of the vector type.
+        $vector:ty,
+        // Type of each individual component, for example `i32`.
+        $scalar:ty,
+        // Names of the components, with parentheses, for example `(x, y)`.
+        ($($components:ident),*),
+        // Name of the operator trait, for example `Add`.
+        $operator:ident,
+        // Name of the function on the operator trait, for example `add`.
+        $func:ident$(,)?
+    ) => {
+        // vector + vector
+        impl std::ops::$operator for $vector {
+            type Output = Self;
+            fn $func(mut self, rhs: $vector) -> Self::Output {
+                $(
+                    self.$components = self.$components.$func(rhs.$components);
+                )*
+                self
+            }
+        }
+
+        // vector + scalar
+        impl std::ops::$operator<$scalar> for $vector {
+            type Output = Self;
+            fn $func(mut self, rhs: $scalar) -> Self::Output {
+                $(
+                    self.$components = self.$components.$func(rhs);
+                )*
+                self
+            }
+        }
+
+        // scalar + vector
+        impl std::ops::$operator<$vector> for $scalar {
+            type Output = $vector;
+            fn $func(self, mut rhs: $vector) -> Self::Output {
+                $(
+                    rhs.$components = rhs.$components.$func(self);
+                )*
+                rhs
+            }
+        }
```

Comment:
These operations make sense:
```
vector * vector     component-wise multiplication
vector * scalar     scaling
scalar * vector     scaling

vector / vector     component-wise division
vector / scalar     scaling

vector + vector     addition
vector - vector     subtraction
```
But these don't:
```
scalar / vector

scalar + vector
vector + scalar

scalar - vector
vector - scalar
```
Correspondingly for the `*Assign` variants.
So we cannot just blanket-implement all 3 combinations, but need to do it depending on the operator.

Furthermore, I would **not** provide `Rem` (`%`) for now, as Godot doesn't have it and semantics would need to be clarified.

Reply:
Godot itself implements these seemingly useless ones too! The only one I added is `%` for float vectors; it actually does have its uses (wrap a vector around in a box). Semantics for negative values etc. are (I assume) properly defined by Rust.

Reply:
> Godot itself implements these seemingly useless ones too!

Well it certainly [doesn't document them](https://docs.godotengine.org/en/latest/classes/class_vector3.html#operators). Which is reason enough for me to exclude them.

Having fewer operators prevents real mistakes like:
```rs
vector += speed * dt;
// both speed and dt are f32
// BAM
```

I'd rather have esoteric operations a bit more verbose if we can catch errors like those. If things get truly painful, users will complain.

Reply:
Regarding the macro, easiest would probably mean if each macro invocation would just define one operator, not 2 or 3 at once. Then we can be very selective on the allowed operand combinations, and it's also immediately visible which ones are supported.

Reply:
Sorry, my mistake. They don't actually exist in the API either. I split the macros up and nuked the stupid ones. The calls are a bit verbose now, but the implementation is as simple as it gets.

I'm itching to implement a macro that lets me literally write its arguments like `vector *= scalar`, but I'll resist that temptation :)